### PR TITLE
Issue 1350: PravegaRequestProcessor fixes, etc.

### LINKED ISSUE #1350: AbstractThreadPoolService.doStop does not play well with InlineExecutor
**Problem description**
If AbstractThreadPoolService.doStop is invoked and its executor is an InlineExecutor it can block on itself.

**Problem location**
```
java.util.concurrent.TimeoutException: null
	at java.util.concurrent.CompletableFuture.timedGet(CompletableFuture.java:1771)
	at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1915)
	at io.pravega.common.concurrent.AbstractThreadPoolService.lambda$1(AbstractThreadPoolService.java:111)
	at io.pravega.common.concurrent.ExecutorServiceHelpers.lambda$0(ExecutorServiceHelpers.java:68)
	at io.pravega.test.common.InlineExecutor.execute(InlineExecutor.java:34)
	at io.pravega.common.concurrent.ExecutorServiceHelpers.execute(ExecutorServiceHelpers.java:66)
	at io.pravega.common.concurrent.AbstractThreadPoolService.doStop(AbstractThreadPoolService.java:103)
	at com.google.common.util.concurrent.AbstractService.stopAsync(AbstractService.java:242)
	at io.pravega.segmentstore.server.containers.StreamSegmentContainer.doStop(StreamSegmentContainer.java:184)
	at io.pravega.segmentstore.server.containers.StreamSegmentContainer.doStop(StreamSegmentContainer.java:168)
	at com.google.common.util.concurrent.AbstractService.stopAsync(AbstractService.java:242)
	at io.pravega.segmentstore.server.store.StreamSegmentContainerRegistry.stopContainer(StreamSegmentContainerRegistry.java:152)
	at io.pravega.segmentstore.server.mocks.LocalSegmentContainerManager.close(LocalSegmentContainerManager.java:86)
	at io.pravega.segmentstore.server.store.ServiceBuilder.closeComponent(ServiceBuilder.java:307)
	at io.pravega.segmentstore.server.store.ServiceBuilder.close(ServiceBuilder.java:136)
	at io.pravega.segmentstore.server.host.handler.PravegaRequestProcessorTest.testSegmentAttribute(PravegaRequestProcessorTest.java:203)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

### PR DESCRIPTION
**Pull request by andreipaduroiu**
**Change log description**
- **AbstractThreadPoolService**: 
    - Rewrote doStop() so it doesn't require an entire extra thread for the duration of shutdown (i.e., made it async). This reduces the overall thread consumption of the application when Container are shut down (since SegmentContainers, and a few sub-components derive from this).
        - This will shorten our build time by about 60 seconds since a few unit tests were hanging (without failure) on shutdown due to insufficient threads being available - PravegaRequestProcessorTest is a prime example.
    - Added Unit tests.
- **AppendProcessor,** **PravegaConnectionListener:** 
    - Small code reorg to meet our current coding guidelines.
-  **PravegaRequestProcessor**:
    - **commitTransaction:** executing seal and merge in parallel (but invoked in this order), which should reduce the latency for this operation. There is no need for merge to wait on seal to complete.
    - Various code reorg to meet our current coding guidelines.
    - Coalesced various calls into shorter blocks and removed unnecessary variables where applicable.

**Purpose of the change**
Fixes #1350.
Reduces the execution time of some unit tests (by some 10 seconds each).

**What the code does**
See description.

**How to verify it**
All existing unit tests must pass. New unit tests must pass too.

### CODE REVIEW

File: common/src/main/java/io/pravega/common/concurrent/FutureHelpers.java [hunk 1]
```diff
@@ -252,6 +252,28 @@
...
 
     /**
+     * Attempts to retrieve the failure cause of a Future. This is equivalent to calling get() on the Future and catching
+     * and unwrapping any exceptions being thrown.
+     *
+     * @param future The Future to check.
+     * @return A Throwable indicating the failure cause, or null if the Future completed normally.
+     * @throws IllegalArgumentException If the Future is not yet completed.
+     */
+    public static Throwable getFailureCause(CompletableFuture<?> future) {
```
Comment (tkaitchuck) [inline common/src/main/java/io/pravega/common/concurrent/FutureHelpers.java]: I currently have a PR which adds a very similar method:
https://github.com/pravega/pravega/pull/1764/files#diff-d89cb9f56cd6e22054026ff28b722371R120

We should reconcile these down to one version. They are semantically different in what they do if the future has completed successfully. I think this should be treated the same as though it was still running: If there is no exception, it should return null. If you move you're method up to just below isSuccessful on line 120 above, it will force a merge conflict for us to resolve reguardless of which PR goes in first.

File: common/src/test/java/io/pravega/common/concurrent/AbstractThreadPoolServiceTests.java [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+import org.junit.rules.Timeout;
+
+/**
+ * Unit tests for the AbstractThreadPoolService class.
+ */
+public class AbstractThreadPoolServiceTests extends ThreadPooledTestSuite {
+    private static final long SHORT_TIMEOUT_MILLIS = 10;
+    private static final Duration TIMEOUT = Duration.ofSeconds(10);
+    @Rule
+    public Timeout globalTimeout = new Timeout(TIMEOUT.getSeconds(), TimeUnit.SECONDS);
```
Comment (tkaitchuck) [inline common/src/test/java/io/pravega/common/concurrent/AbstractThreadPoolServiceTests.java]: final

File: common/src/test/java/io/pravega/common/concurrent/AbstractThreadPoolServiceTests.java [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+        val stopException = new IntentionalException("stop");
+        s.errorHandler(stopException);
+        val runException = new IntentionalException("run");
+        s.runFuture.completeExceptionally(runException);
+
+        AssertExtensions.assertThrows(
+                "Service did not fail.",
+                () -> s.awaitTerminated(),
+                ex -> ex instanceof IllegalStateException);
+        Assert.assertEquals("Unexpected state upon failed shutdown.", Service.State.FAILED, s.state());
```
Comment (tkaitchuck) [inline common/src/test/java/io/pravega/common/concurrent/AbstractThreadPoolServiceTests.java]: You can use static imports for these assert methods and save space.

File: segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/AppendProcessor.java [hunk 4]
```diff
@@ -146,16 +173,16 @@ public void performNextWrite() {
...
-            handleAppendResult(append, e);
-            if (e == null) {
-                WRITE_STREAM_SEGMENT.reportSuccessEvent(timer.getElapsed());
-            } else {
-                WRITE_STREAM_SEGMENT.reportFailEvent(timer.getElapsed());
-            }
-        }).whenComplete((v, e) -> {
-            append.getData().release();  
-        });
+        storeAppend(append)
```
Comment (tkaitchuck) [inline segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/AppendProcessor.java]: Is this reformatting needed?

File: segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java [hunk 5]
```diff
@@ -226,104 +169,96 @@ public void updateSegmentAttribute(UpdateSegmentAttribute updateSegmentAttribute
...
-                } else {
-                    handleException(requestId, segmentName, "Update attribute", e);
-                }
-            }
-        }).exceptionally((Throwable e) -> {
-            handleException(requestId, segmentName, "Update attribute", e);
-            return null;
-        });
+        segmentStore
+                .updateAttributes(segmentName, Collections.singletonList(update), TIMEOUT)
```
Comment (tkaitchuck) [inline segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java]: This really shouldn't line wrap here.

File: segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java [hunk 6]
```diff
@@ -226,104 +169,96 @@ public void updateSegmentAttribute(UpdateSegmentAttribute updateSegmentAttribute
...
-            return null;
-        }).exceptionally((Throwable e) -> {
-            handleException(getStreamSegmentInfo.getRequestId(), segmentName, "Get segment info", e);
-            return null;
-        });
+        segmentStore.getStreamSegmentInfo(segmentName, false, TIMEOUT)
+                .thenAccept(properties -> {
+                    if (properties != null) {
+                        StreamSegmentInfo result = new StreamSegmentInfo(getStreamSegmentInfo.getRequestId(),
+                                properties.getName(),
```
Comment (tkaitchuck) [inline segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java]: If you're going to format something with one item per line, than indent on column so it's clear where they go.

Where I've ended up with my settings:
I've settled on function arguments being wrapped as needed but indented on column and chained invocations going on each on their own line if wrapping is needed at all. In both cases using column indent. I always use column indent except for constructor args, where I find they end up way off to the end of the line, so I use default indentation and wrap as needed. 
(As you can see this file was written before I settled on these settings)

File: segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java [hunk 7]
```diff
@@ -226,104 +169,96 @@ public void updateSegmentAttribute(UpdateSegmentAttribute updateSegmentAttribute
...
-            return null;
-        }).exceptionally((Throwable e) -> {
-            handleException(request.getRequestId(), transactionName, "Get transaction info", e);
-            return null;
-        });
+        segmentStore.getStreamSegmentInfo(transactionName, false, TIMEOUT)
+                .thenAccept(properties -> {
+                    if (properties != null) {
+                        TransactionInfo result = new TransactionInfo(request.getRequestId(),
+                                request.getSegment(),
```
Comment (tkaitchuck) [inline segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java]: see my above comment

File: common/src/main/java/io/pravega/common/concurrent/FutureHelpers.java (hunk 1 above)
Comment (andreipaduroiu) [inline common/src/main/java/io/pravega/common/concurrent/FutureHelpers.java]: Yes. I am going to use your method.

File: segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/AppendProcessor.java (hunk 4 above)
Comment (andreipaduroiu) [inline segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/AppendProcessor.java]: I found it pretty hard to follow the flow here since some callbacks are on their own lines and some are not. I standardized them all to start on new lines, so that the IDE can properly format them.

File: segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java (hunk 5 above)
Comment (andreipaduroiu) [inline segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java]: FIxed

File: segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java (hunk 6 above)
Comment (andreipaduroiu) [inline segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java]: Fixed

File: segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java [hunk 8]
```diff
@@ -377,101 +376,89 @@ private void handleException(long requestId, String segment, String operation, T
...
                             return null;
-                        });
-                        return null;
-                    })
-                    .exceptionally((Throwable e) -> {
-                        handleException(requestId, transactionName, "Commit transaction", e);
-                        return null;
-                    });
+                        }));
+        val merge = segmentStore
```
Comment (shiveshr) [inline segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java]: what could be the consequence of doing thesmerge and seal in parallel?

for example, a bad application code could be writing to a txn while calling commit on it concurrently.. 

if we are merging an unsealed segment, do we run the risk of some unwanted races?

Comment (andreipaduroiu) [reply segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java]: You cannot merge an unsealed segment. The OperationProcessor will reject a merge if the segment isn't sealed. If you look, this is all server-side code and is invoked every time someone wants to merge a txn. On the SSS side, we could choose to automatically seal the segment upon a merge, but we left it open, which gives more flexibility. This change has no effect from the outside, except that it will execute faster.
